### ISSUE TITLE
Responsive Spaceline

### ISSUE THREAD

**Issue by domtronn** [TheBB/spaceline#146]
Hi there, thanks for such a great package!

I've just started using the version that includes the responsive spaceline changes, however, is there a way to disable this feature? And is there any documentation on the `:priority` property and how it works?

I'm getting segments with important information disappear in favour of less useful ones.

Also, if I have two _blocks_ of segments, e.g.

```el
     ...
     ((segment-1
       semgent-2)
      :face highlight-face
      :separator "/")

      segment-3
     ...
```

Is there a way to have _both_ blocks of segments disappear at the _same time_?

**Comment by deb0ch** [comment]
Hi !

The priorities segment by segment haven't been really thought of thoroughly, as for now our efforts have been focused on getting the core algorithm working and optimized.

So if you feel like the order is wrong and should be otherwise, you're probably right, if you do correct some priorities to something that seems more sensible to you, don't hesitate to contribute it :smile_cat: 

As for the `:priority` themselves, they are indeed not documented in the README. The closest to a docs that you can find is in the docstrings, in `spaceline.el` lines 418 and 667. Again, don't hesitate to contribute if you feel like a proper documentation is needed :smile_cat: 

Basically, the `:priority` is a segment attribute like any other (`:face`, `:separator`...), the lower the number the higher the priority, with the lowest (and default) priority being `-1`. This is why, if we didn't assign a priority to the segment that you like, it will automatically have the lowest. Floating point values work, so you can always insert a priority between two others :hamburger:

It is not possible in the current implementation to have several "blocks of segments" disappear at the same time, only segments that are grouped together in a "block" will disappear at the same time. I don't know if nested blocks are supported, @TheBB ?

And I wonder why you would even want that: if there is room to display one block, why not display it ?

**Comment by TheBB** [comment]
Nested segments are supported, but separators only show between top-level segments, and only top-level segments will be shown or hidden, and they do so independently. If you want segments to be shown or hidden together you should bundle them in a combined top-level segment.

**Comment by deb0ch** [comment]
Blocks of segments are supported, but specifically blocks of blocks of segments ? Is the recursion infinite ?

**Comment by TheBB** [comment]
Well, not infinite, but arbitrarily deep. ;-)

**Comment by TheBB** [comment]
https://github.com/TheBB/spaceline/blob/master/tests.el#L21 for example

**Comment by TheBB** [comment]
https://github.com/TheBB/spaceline/blob/master/tests.el#L51 too

**Comment by domtronn** [comment]
Thanks for the quick reply @deb0ch & @TheBB. Yeah I'm currently playing with the priorities to try to get it the way I want, so thanks for the useful information - I'll definitely contribute once I've got my head around it 👍 

I'm the maintainer of [`spaceline-all-the-icons`](https://github.com/domtronn/spaceline-all-the-icons.el), a theme for **Spaceline** using font icons, one of the reasons for creating it was to use a font icon for the block separators _(`powerline-default-separator` the slants, arrows waves etc)_ so that it'll work with Mac Retina **sRGB** displays, _(below screenshot shows the issue with XPM images)_

<img width="1020" alt="screen shot 2017-05-30 at 11 34 25" src="https://cloud.githubusercontent.com/assets/5881414/26579525/1db39d1e-452c-11e7-8d87-c6f87841f2c6.png">

And so I can set the character's foreground and background rather than use XPM images. The problem is that those colours aren't dynamic since they need to know the _previous_ & _next_ top level segment background colours to render the transition correctly.

I'm currently doing [this](https://github.com/domtronn/spaceline-all-the-icons.el/blob/master/spaceline-all-the-icons.el#L154-L165) and it works nicely, but when blocks disappear from the responsiveness, the transition colouring is wrong because the next segments aren't there anymore! 🙁  _(see below screenshot)_

<img width="472" alt="screen shot 2017-05-30 at 09 26 30" src="https://cloud.githubusercontent.com/assets/5881414/26579815/68088a72-452d-11e7-8bd6-e4897b6abf44.png">

Also, I've tried placing the separators inside the top level segments but I can't get them to be `:tight` so they have padding either side which is a fixed background colour - which always clashes with _one_ of the colours I'm transitioning between - hopefully that makes sense! 😅

**tl;dr**

I guess my best bet is to try and make the segments that render the separators dynamic somehow, and programatically calculate their foreground/background colour?

Is there a way to get the next segment colour?

**Comment by TheBB** [comment]
Ah okay, sorry for messing up your package. I wasn't aware you were doing this.

Is there a reason you're hand-coding your separators instead of using

```elisp
(setq powerline-default-separator 'utf-8
      powerline-utf-8-separator-left ...)
```

I presume there is…

> Is there a way to get the next segment colour?

Yes, segment code runs in an environment where `default-face` and `other-face` alternate. But if your separators are themselves segments, I don't know how you deal with that.

Maybe the best option here is to add some features to spaceline so that the separators (can) work the way you want them to. Some sort of `(spaceline|define-separator ...)` macro.

**Comment by TheBB** [comment]
I don't think that should be too hard. Powerline separators are just functions that take two faces as arguments and spit out some kind of image or string. Currently we have hardcoded this function to `<powerline>-<separator>-<left/right>` but we can make that more flexible. Then you can plug into that interface.

**Comment by jakemcarthur** [comment]
The original question was "is there a way to disable this feature?", and I also am curious what the answer is.

**Comment by deb0ch** [comment]
There is for now no way to disable this feature other than cloning the repo locally and checking out an older version.

The reason is that this feature required a deep change in the architecture of the Spaceline algorithm, so disabling it would have very close to no benefits on a performance standpoint.

So I guess it would be possible to disable this feature, but I can't grasp in which way doing so would be useful ?

**Comment by jakemcarthur** [comment]
I just want to avoid bugs like #147 and avoid another annoyances due to the priorities not being good, until this feature is in a more stable shape.

**Comment by deb0ch** [comment]
The simplest if you want to deactivate the feature is to go to [spaceline.el:482](https://github.com/TheBB/spaceline/blob/master/spaceline.el#L482) and comment out the `when` expression.

**Comment by domtronn** [comment]
Thanks for that snippet! That does disable the responsive behaviour. 🙂 

But if that's the case, could you not allow users to opt into this feature with something like the following?

```el
(defcustom spaceline-responsive-display t
  "Adjust the visibility of segments based on priority and window width."
  :group spaceline
  :type 'boolean)

;; ...
(when (or (not spaceline-responsive-display)
          (spaceline--adjust-to-window ,priority-symbol fmt))
   (setq fmt (spaceline--render-mode-line ,left-code ,right-code)))
;; ...
```

For me personally, I just don't understand why you would ever _opt_ to hide useful information, as they say, _"better to have it and not need it, than to need it and not have it"_ 😅

**Comment by TheBB** [comment]
Sure, we can add such a feature.

> I just don't understand why you would ever opt to hide useful information

We don't. If the window is too narrow, useful information will be hidden, period. This is about controlling what is hidden and what is not.

**Comment by deb0ch** [comment]
Indeed, for @jakemcarthur if it is just about the priorities not being good I would suggest that fixing the priorities would be an approach worth considering.

Again, that's a point where more work is still needed, si it would be even nicer if improved priorities were contributed :smile_cat:

**Comment by TheBB** [comment]
Here: https://github.com/TheBB/spaceline/commit/b1c41ddd290153fcc1df8f1d0fb54062e9943894

**Comment by deb0ch** [comment]
Though after a while, when all the bugs have been flattened out, I would suggest removing that option.

I think that too many configuration variables makes configuration cumbersome. When there is no bugs anymore with that feature (let us say arbitrarily, in a year or two to leave other packages the time to adapt ?) there will be no point having that option, right ?

**Comment by jakemcarthur** [comment]
I have no objection to this variable being ignored once the kinks have been ironed out.

**Comment by deb0ch** [comment]
@domtronn Hi, would you please mind to opening a separate issue with what you need to make Spaceline and All The Icons work well together again ?

I don't mind getting to implementing it, but the report here is cluttered by the rest of the discussion and it is not exactly clear to me what you need, more details would be nice.

**Comment by deb0ch** [comment]
Yay 😀

**Comment by apnewberry** [comment]
I'm trying to get the buffer name to have priority so that it's always there whenever possible (I want buffer name to be the last thing to disappear). I'm looking around `spaceline-config.el` but I don't see `(buffer-name)` like I expected. How can I set the priority of buffer name so it doesn't disappear until last? This kind of question might be addressed more generally by #183.
